### PR TITLE
Flexible gradients normalization

### PR DESCRIPTION
**Pull request by VaKonS**
Since results with regular and normalized gradients are quite different, maybe it would be useful to apply floating point normalization factor instead of boolean "on/off" option.

Difference – smooth choice between ordinary and normalized gradients _(above are the old option values, below are the new ones)_:

flexible_gradients_normalization_full_range

**Disadvantages**: new syntax is incompatible with old option (need to use "-normalize_gradients **1.0**").
 
New "normalize_gradients" value "0.0" (or skipped option) corresponds to old value "false" (normalization skipped completely), and new value "1.0" corresponds to old value "true" (old code is used unchanged).

***Update**: added normalization auto-disabling when content losses are low – it helps to begin stylization faster.
With this, option values above 1.0 always make noisy garbage, therefore its range _should_ be limited to 1.

### CODE REVIEW

**Comment by ProGamerGov** [comment]
Is there any way to have `-normalize_gradients 1.0` and `-normalize_gradients 1` both be acceptable inputs, like the `-learning_rate` command?

**Comment by VaKonS** [comment]
**ProGamerGov**, both "-normalize_gradients 1.0" and "-normalize_gradients 1" should work, the value is compared as a number.

With old boolean switch you had to specify "-normalize_gradients" or nothing, if I'm not mistaken.
With floating point it's necessary to use "-normalize_gradients _n_", or skip switch too.

It's possible to add second switch, "-ng_power _n_" for example, and leave old switch as is.
But then "-normalize_gradients" will duplicate "-ng_power 1" function, does it make sense?

**Comment by ProGamerGov** [comment]
The [effect of different `-normalize_gradients` values on loss graphs](https://github.com/ProGamerGov/Neural-Graph/wiki/Graphs), is interesting:

**Comment by ProGamerGov** [comment]
Using a value higher than 1 reminds me of what happens when you replace the 1e-8 from [this line of code](https://github.com/jcjohnson/neural-style/blob/e1ff8fa45dded440ede9e6cc084c3334e3859f67/neural_style.lua#L556), with a larger number.

**Comment by htoyryla** [comment]
I am wondering what this partial normalization actually amounts to. It seems to that it is not normalization at all but simply a form of scaling the gradients. And when we scale the gradients, aren't we indirectly affecting the learning rate (smaller gradient => smaller steps => slower learning but more stable). 

Another similar phenomenon is increasing or decreasing both content and style weight so that their ratio stays the same. Many of us know that doing that gives different results. Higher absolute weights results in higher loss which give higher gradients which result in larger changes per iteration (similar to higher learning rate).

**Comment by VaKonS** [comment]
**ProGamerGov**, regarding optimization of parameters - I'm thinking about "self-regulated" normalization, based on losses, but didn't figured out yet how to make it.

**htoyryla**, it is scaling, but scaling of previously _normalized_ gradients, so it's something like image-dependent scaling, I guess.

By the way, during tests I've probably found one thing (see second commit in this pull request): sometimes it's useful to temporarily turning gradients normalization off.

**Comment by htoyryla** [comment]
"but scaling of previously normalized gradients, so it's something like image-dependent scaling"

I see the point. Still I wonder if this is not simply another setting for learning rate.

"By the way, during tests I've probably found one thing (see second commit in this pull request): sometimes it's useful to temporarily turning gradients normalization off."

Somehow I think this too is actually a learning rate issue: sometimes you need a larger learning rate initially to get the learning started, especially when the gradients are very small (which can happen with normalization).

**Comment by htoyryla** [comment]
"I'm thinking about "self-regulated" normalization, based on losses, but didn't figured out yet how to make it."

Someone experimented once on adjusting style weights based on losses, starting from my modification for layer-specific style weights. It is in the issues somewhere, I'll post the link if I find it.

**Comment by htoyryla** [comment]
The discussion about self-regulated weights is in this thread https://github.com/jcjohnson/neural-style/issues/237 . It is a long one, there are several threads going on, the relevant one is between me and bmaltais.

**Comment by VaKonS** [comment]
**htoyryla**, if this normalization scaling is an "alias" for lerning rate adjustment, then maybe it's possible to "self-regulate" learning rate, based on losses?
It could help with cases when optimization stucks.

**Comment by htoyryla** [comment]
Never mind the terminology, I was looking for the rationale for scaling the gradient, as well as how it affects the optimization process. Neural-style does not, however, pass the learning_rate flag to l-bfgs, only adam; I am not aware what mechanisms l-bfgs uses internally for learning rate.

**Comment by VaKonS** [comment]
**htoyryla**, in this case (if "learning rate" doesn't affect l-bfgs), this pull request probably makes sense, because it works with both optimizers.

**Comment by VaKonS** [comment]
**ProGamerGov**, now I've limited the range to 1, because with auto-skipped normalization values higher than 1.0 stably make useless noise after 4th iteration. Factors <= 1 work as expected.

**Comment by VaKonS** [comment]
The commit [7682e2e](https://github.com/jcjohnson/neural-style/pull/374/commits/7682e2e24b650206afeddc576a6ca0778d11260c) is not related to this request.
I'll remove it as soon as will find out how to make it in browser.
Sorry.

**Comment by VaKonS** [comment]
Since there is no @jcjohnson's opinion, I guess it means that this functionality is not needed, at least in base project, so I'm closing the request.

Anyway, thanks to all participants.
